TLM/TLM/Util/Extensions/NetSegmentExtensions.cs

```diff
@@ -9,5 +53,130 @@ public static class NetSegmentExtensions {
             netSegment.m_flags.CheckFlags(
                 required: NetSegment.Flags.Created,
                 forbidden: NetSegment.Flags.Collapsed | NetSegment.Flags.Deleted);
+
+        public static NetInfo.Lane GetLaneInfo(this ref NetSegment netSegment, int laneIndex) =>
+            netSegment.Info?.m_lanes?[laneIndex];
+
+        public static GetSegmentLaneIdsEnumerable GetSegmentLaneIdsAndLaneIndexes(this ref NetSegment netSegment) {
+            NetInfo netInfo = netSegment.Info;
+            uint initialLaneId = netSegment.m_lanes;
+            NetLane[] laneBuffer = NetManager.instance.m_lanes.m_buffer;
+            if (netInfo == null) {
+                return new GetSegmentLaneIdsEnumerable(0, 0, laneBuffer);
+            }
+
+            return new GetSegmentLaneIdsEnumerable(initialLaneId, netInfo.m_lanes.Length, laneBuffer);
+        }
+
+        /// <summary>
+        /// Assembles a geometrically sorted list of lanes for the given segment.
+        /// If the <paramref name="startNode"/> parameter is set only lanes supporting traffic to flow towards the given node are added to the list, otherwise all matched lanes are added.
+        /// </summary>
+        /// <param name="segmentId">segment id</param>
+        /// <param name="netSegment">segment data</param>
+        /// <param name="startNode">reference node (optional)</param>
+        /// <param name="laneTypeFilter">lane type filter, lanes must match this filter mask</param>
+        /// <param name="vehicleTypeFilter">vehicle type filter, lanes must match this filter mask</param>
+        /// <param name="reverse">if true, lanes are ordered from right to left (relative to the
+        ///     segment's start node / the given node), otherwise from left to right</param>
+        /// <param name="sort">if false, no sorting takes place
+        ///     regardless of <paramref name="reverse"/></param>
+        /// <returns>sorted list of lanes for the given segment</returns>
+        public static IList<LanePos> GetSortedLanes(
+            this ref NetSegment netSegment,
+            bool? startNode,
+            NetInfo.LaneType? laneTypeFilter = null,
+            VehicleInfo.VehicleType? vehicleTypeFilter = null,
+            bool reverse = false,
+            bool sort = true) {
+            // TODO refactor together with getSegmentNumVehicleLanes, especially the vehicle type and lane type checks
+            NetManager netManager = Singleton<NetManager>.instance;
+            var laneList = new List<LanePos>();
```

Comment:
Pre-size to num lanes and then compact before returning? (I can't rememeber how many elements list allocates by default so probably no need but thought I'd ask just in case)